File: app/src/main/java/com/babestudios/companyinfouk/ui/chargesdetails/ChargesDetailsAdapter.java
package com.babestudios.companyinfouk.ui.chargesdetails;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.babestudios.companyinfouk.R;
import com.babestudios.companyinfouk.data.model.charges.ChargesItem;

import butterknife.BindView;
import butterknife.ButterKnife;

class ChargesDetailsAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

	private static final int TYPE_HEADER = 0;
	private static final int TYPE_TRANSACTION = 1;

	private ChargesItem chargesItem;

	Context context;

	ChargesDetailsAdapter(Context c, ChargesItem chargesItem) {
		this.chargesItem = chargesItem;
		context = c;
	}

	@Override
	public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
		View itemLayoutView;
		if (viewType == TYPE_TRANSACTION) {
			itemLayoutView = LayoutInflater.from(parent.getContext())
					.inflate(R.layout.charges_details_transactions_list_item, parent, false);

			return new TransactionsViewHolder(itemLayoutView);
		} else if (viewType == TYPE_HEADER) {
			itemLayoutView = LayoutInflater.from(parent.getContext())
					.inflate(R.layout.charges_details_header_item, parent, false);

			return new HeaderViewHolder(itemLayoutView);
		}
		throw new RuntimeException("There is no type that matches the type " + viewType + " + make sure your using types correctly");
	}

	@Override
	public void onBindViewHolder(RecyclerView.ViewHolder viewHolder, int position) {
		if (viewHolder instanceof HeaderViewHolder) {
			((HeaderViewHolder) viewHolder).textViewDeliveredOn.setText(chargesItem.deliveredOn);
			((HeaderViewHolder) viewHolder).textViewStatus.setText(chargesItem.status);
			((HeaderViewHolder) viewHolder).textViewContainsFloatingCharge.setText(chargesItem.particulars.containsFloatingCharge ? "YES" : "NO");
			((HeaderViewHolder) viewHolder).textViewFloatingChargeCoversAll.setText(chargesItem.particulars.floatingChargeCoversAll ? "YES" : "NO");
			((HeaderViewHolder) viewHolder).textViewContainsNegativePledge.setText(chargesItem.particulars.containsNegativePledge ? "YES" : "NO");
			((HeaderViewHolder) viewHolder).textViewContainsFixedCharge.setText(chargesItem.particulars.containsFixedCharge ? "YES" : "NO");
			if(chargesItem.satisfiedOn == null) {
				((HeaderViewHolder) viewHolder).textViewLabelSatisfiedOn.setVisibility(View.GONE);
				((HeaderViewHolder) viewHolder).textViewSatisfiedOn.setVisibility(View.GONE);
			}else{
				((HeaderViewHolder) viewHolder).textViewSatisfiedOn.setVisibility(View.VISIBLE);
				((HeaderViewHolder) viewHolder).textViewLabelSatisfiedOn.setVisibility(View.VISIBLE);
				((HeaderViewHolder) viewHolder).textViewSatisfiedOn.setText(chargesItem.satisfiedOn);
			}
			((HeaderViewHolder) viewHolder).textViewPersonsEntitled.setText(chargesItem.personsEntitled.get(0).name);

		} else if (viewHolder instanceof TransactionsViewHolder) {
			((TransactionsViewHolder) viewHolder).textViewFilingType.setText(chargesItem.transactions.get(position - 1).filingType);
			((TransactionsViewHolder) viewHolder).textViewDeliveredOn.setText(chargesItem.transactions.get(position - 1).deliveredOn);
		}

	}

	@Override
	public int getItemViewType(int position) {
		if (isPositionHeader(position)) {
			return TYPE_HEADER;
		} else {
			return TYPE_TRANSACTION;
		}
	}

	private boolean isPositionHeader(int position) {
		return position == 0;
	}

	public long getItemId(int position) {
		return position;
	}

	@Override
	public int getItemCount() {
		return chargesItem.transactions.size() + 1;
	}

	public class HeaderViewHolder extends RecyclerView.ViewHolder {
		@BindView(R.id.textViewDeliveredOn)
		TextView textViewDeliveredOn;
		@BindView(R.id.textViewStatus)
		TextView textViewStatus;
		@BindView(R.id.textViewContainsFloatingCharge)
		TextView textViewContainsFloatingCharge;
		@BindView(R.id.textViewFloatingChargeCoversAll)
		TextView textViewFloatingChargeCoversAll;
		@BindView(R.id.textViewContainsNegativePledge)
		TextView textViewContainsNegativePledge;
		@BindView(R.id.textViewContainsFixedCharge)
		TextView textViewContainsFixedCharge;
		@BindView(R.id.textViewLabelSatisfiedOn)
		TextView textViewLabelSatisfiedOn;
		@BindView(R.id.textViewSatisfiedOn)
		TextView textViewSatisfiedOn;
		@BindView(R.id.textViewPersonsEntitled)
		TextView textViewPersonsEntitled;

		HeaderViewHolder(View itemView) {
			super(itemView);
			ButterKnife.bind(this, itemView);
		}
	}

	public class TransactionsViewHolder extends RecyclerView.ViewHolder {

		@BindView(R.id.textViewFilingType)
		TextView textViewFilingType;
		@BindView(R.id.textViewDeliveredOn)
		TextView textViewDeliveredOn;


		TransactionsViewHolder(View itemView) {
			super(itemView);
			ButterKnife.bind(this, itemView);
		}
	}
}

File: app/src/main/java/com/babestudios/companyinfouk/ui/chargesdetails/ChargesDetailsAdapter.kt
package com.babestudios.companyinfouk.ui.chargesdetails

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import com.babestudios.companyinfouk.R
import com.babestudios.companyinfouk.data.model.charges.ChargesItem

import butterknife.BindView
import butterknife.ButterKnife

internal class ChargesDetailsAdapter(var context: Context, private val chargesItem: ChargesItem) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

	override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
		val itemLayoutView: View
		if (viewType == TYPE_TRANSACTION) {
			itemLayoutView = LayoutInflater.from(parent.context)
					.inflate(R.layout.charges_details_transactions_list_item, parent, false)

			return TransactionsViewHolder(itemLayoutView)
		} else if (viewType == TYPE_HEADER) {
			itemLayoutView = LayoutInflater.from(parent.context)
					.inflate(R.layout.charges_details_header_item, parent, false)

			return HeaderViewHolder(itemLayoutView)
		}
		throw RuntimeException("There is no type that matches the type $viewType + make sure your using types correctly")
	}

	override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) {
		if (viewHolder is HeaderViewHolder) {
			viewHolder.textViewDeliveredOn!!.text = chargesItem.deliveredOn
			viewHolder.textViewStatus!!.text = chargesItem.status
			viewHolder.textViewContainsFloatingCharge!!.text = if (chargesItem.particulars.containsFloatingCharge) "YES" else "NO"
			viewHolder.textViewFloatingChargeCoversAll!!.text = if (chargesItem.particulars.floatingChargeCoversAll) "YES" else "NO"
			viewHolder.textViewContainsNegativePledge!!.text = if (chargesItem.particulars.containsNegativePledge) "YES" else "NO"
			viewHolder.textViewContainsFixedCharge!!.text = if (chargesItem.particulars.containsFixedCharge) "YES" else "NO"
			if (chargesItem.satisfiedOn == null) {
				viewHolder.textViewLabelSatisfiedOn!!.visibility = View.GONE
				viewHolder.textViewSatisfiedOn!!.visibility = View.GONE
			} else {
				viewHolder.textViewSatisfiedOn!!.visibility = View.VISIBLE
				viewHolder.textViewLabelSatisfiedOn!!.visibility = View.VISIBLE
				viewHolder.textViewSatisfiedOn!!.text = chargesItem.satisfiedOn
			}
			viewHolder.textViewPersonsEntitled!!.text = chargesItem.personsEntitled[0].name

		} else if (viewHolder is TransactionsViewHolder) {
			viewHolder.textViewFilingType!!.text = chargesItem.transactions[position - 1].filingType
			viewHolder.textViewDeliveredOn!!.text = chargesItem.transactions[position - 1].deliveredOn
		}

	}

	override fun getItemViewType(position: Int): Int {
		return if (isPositionHeader(position)) {
			TYPE_HEADER
		} else {
			TYPE_TRANSACTION
		}
	}

	private fun isPositionHeader(position: Int): Boolean {
		return position == 0
	}

	override fun getItemId(position: Int): Long {
		return position.toLong()
	}

	override fun getItemCount(): Int {
		return chargesItem.transactions.size + 1
	}

	inner class HeaderViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView) {
		@JvmField
		@BindView(R.id.textViewDeliveredOn)
		internal var textViewDeliveredOn: TextView? = null
		@JvmField
		@BindView(R.id.textViewStatus)
		internal var textViewStatus: TextView? = null
		@JvmField
		@BindView(R.id.textViewContainsFloatingCharge)
		internal var textViewContainsFloatingCharge: TextView? = null
		@JvmField
		@BindView(R.id.textViewFloatingChargeCoversAll)
		internal var textViewFloatingChargeCoversAll: TextView? = null
		@JvmField
		@BindView(R.id.textViewContainsNegativePledge)
		internal var textViewContainsNegativePledge: TextView? = null
		@JvmField
		@BindView(R.id.textViewContainsFixedCharge)
		internal var textViewContainsFixedCharge: TextView? = null
		@JvmField
		@BindView(R.id.textViewLabelSatisfiedOn)
		internal var textViewLabelSatisfiedOn: TextView? = null
		@JvmField
		@BindView(R.id.textViewSatisfiedOn)
		internal var textViewSatisfiedOn: TextView? = null
		@JvmField
		@BindView(R.id.textViewPersonsEntitled)
		internal var textViewPersonsEntitled: TextView? = null

		init {
			ButterKnife.bind(this, itemView)
		}
	}

	inner class TransactionsViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView) {

		@JvmField
		@BindView(R.id.textViewFilingType)
		internal var textViewFilingType: TextView? = null
		@JvmField
		@BindView(R.id.textViewDeliveredOn)
		internal var textViewDeliveredOn: TextView? = null


		init {
			ButterKnife.bind(this, itemView)
		}
	}

	companion object {

		private val TYPE_HEADER = 0
		private val TYPE_TRANSACTION = 1
	}
}
